vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,364 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/csv_parser/parse_example_csv.h"
+
+#include "vw/core/best_constant.h"
+#include "vw/core/parse_primitives.h"
+#include "vw/core/parser.h"
+
+#include <string>
+
+namespace VW
+{
+namespace parsers
+{
+namespace csv
+{
+int csv_to_examples(VW::workspace* all, io_buf& buf, VW::multi_ex& examples)
+{
+  return all->csv_converter->parse_csv(all, buf, examples[0]);
+}
+
+int parser::parse_csv(VW::workspace* all, io_buf& buf, VW::example* ae)
+{
+  bool first_read = false;
+  if (_header_fn.empty()) { first_read = true; }
+  // This function consumes input until it reaches a '\n' then it walks back the '\n' and '\r' if it exists.
+  size_t num_bytes_consumed = read_line(all, ae, buf);
+  // Read the data if header exists.
+  if (!all->csv_no_header && first_read) { num_bytes_consumed += read_line(all, ae, buf); }
```

Comment:
Currently we just consider that the headers exists in the first line by default. While feature values can also be strings, which cause difficulty to recognise them correctly, maybe we can detect headers by looking into the label columns, as labels columns values are likely to be numbers, and also multiclass label name strings are always specified. But of course it's not always reliable when it comes to prediction as there would be no label then.

For the multiple input files, I guess you mean the feature that is in progress here: https://github.com/VowpalWabbit/vowpal_wabbit/pull/3891 Will look into that if it gets merged. Maybe we have to reset some marker variables in the parser to parse multiple input files, because they may not have the same structure between files.

Reply:
Just realized that we also can't rely on the label columns as label types also vary, it can also be a string with special meanings.

Then I just checked how Pandas handle this. https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html They support passing the column names through the options, and infer based on that. Maybe later we can use the same way to get the multiple input files stuff working.

Reply:
Just changed the `--csv_no_header` options into `--csv_header`. We can now Input the feature names (and/or combined with namespaces using the --csv_ns_separator) separated with ','. By default, if just input value '1', CSV files are assumed to have a header with feature and/or namespaces names in the CSV first line. While with value '0', CSV files are assumed to have no header. If a string is inputted, we will auto-detect the header from CSV first line and get rid of it.